Add proxy peering support to trusted clusters

dboslee opened the pull request:
Adds proxy peering support to trusted clusters. This is done using the`types.TunnelConnections` reported by each proxy that contains which remote cluster is connected to which proxy.

This removes the concept of `clusterPeers` and merges the `clusterPeer` functionality into the `remoteSite` struct.

When proxy peering is enabled the remote site is checked first for local tunnels and then peer proxies are attempted based on the `types.TunnelConnections` available for that site.

Support for dialing remote auth servers through peer proxies was added. This was accomplished by adding a `DialAuth bool` to the grpc DialRequest that indicates the dialer wants to be connected to the remote sites auth server.

When proxy recording is enabled remote sites will record at the agent side of the connection in order to have direct access to the underlying ssh connection.

A new ssh channel was added in order for agents to retrieve the `types.ClusterNetworkingConfig` of the cluster they are connecting to. This was necessary as agents do not have an auth client to the trusted cluster and we need to determine if proxy peering is enabled.

dboslee committed:
Add proxy peering support to trusted clusters

dboslee committed:
Satisfy the linter

dboslee committed:
Use find api call instead of ping

dboslee committed:
Clone tunnel to prevent data race

dboslee committed:
Merge branch 'master' into david/reversetunnel/remote-site

dboslee committed:
Return cached upgrade bool on failure

espadolini commented on lib/reversetunnel/agent.go:
```diff
@@ -592,6 +595,36 @@ func (a *agent) handleDrainChannels() error {
+	ctx, cancel = context.WithTimeout(ctx, defaults.DefaultIOTimeout)
+	defer cancel()
+
+	channel, requests, err := a.client.OpenChannel(ctx, chanNetworkConfig, nil)
```
Why isn't this a global request?

dboslee replied:
Good idea, I got confused by the difference in behavior between global and channel requests and thought you couldn't send a reply payload to either. But turns out you can for global requests and just not channel requests. I will swap it out for a global request.

dboslee replied:
Switched to global requests here [bb4b92b](https://github.com/gravitational/teleport/pull/24702/commits/bb4b92ba5056e3277f13b8a5f63f31d405b34e98)

dboslee replied:
Whoops the reverse tunnel server needed its own global request handler [6563642](https://github.com/gravitational/teleport/pull/24702/commits/6563642a30d9528e66081973ac77ffc028654773)

espadolini replied:
The fact that we've been silently swallowing global requests (completely screwing over any potential for sensible extensibility of the protocol) means that we can't use them reliably unless we know that the server at the other end is sane. 😢

It seems like the only "safe" thing we can do from the client side of the reverse tunnel is open a channel (which, when unknown, will get correctly rejected). What do you think about attempting to open a channel with a name like `reverse-tunnel-v2@goteleport.com` at the beginning of the connection and expecting it to be rejected with a custom rejection code to signify that the server will actually follow the SSH protocol (by rejecting unknown requests and handling requests that want a reply)?

dboslee replied:
@espadolini Let me put something together 😄

espadolini commented on lib/reversetunnel/agent.go:
```diff
@@ -659,10 +692,12 @@ func (a *agent) handleDiscovery(ch ssh.Channel, reqC <-chan *ssh.Request) {
+	chanDiscovery        = "teleport-discovery"
+	chanDiscoveryReq     = "discovery"
+	chanNetworkConfig    = "teleport-netconfig"
+	chanNetworkConfigReq = "netconfig"
```
This should be a global request rather than a channel, but either way:
```suggestion
	chanNetworkConfigReq = "netconfig@goteleport.com"
```

espadolini reviewed (commented):
Just a quick drive-by, not a full review

dboslee committed:
Switch to using ssh global requests for netconfig requests

dboslee committed:
Merge remote-tracking branch 'origin/master' into david/reversetunnel/remote-site

dboslee committed:
Move dial auth request into its own frame

dboslee commented on api/proto/teleport/legacy/client/proto/proxyservice.proto:
```diff
@@ -46,6 +46,11 @@ message DialRequest {
+  // when trying to dial a clusters auth server.
+  string ClusterName = 5;
+  // DialAuth if set will make a dial request to the given clusters auth service.
+  bool DialAuth = 6;
```
Moved it into its own frame here [1c65ecb](https://github.com/gravitational/teleport/pull/24702/commits/1c65ecb9c1ef81fb43bf4ad35fbcd9ae865c6ce5)

dboslee committed:
Simply error handling based on code review

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

dboslee commented on lib/proxy/peer/service.go:
```diff
@@ -40,28 +40,38 @@ func (s *proxyService) DialNode(stream proto.ProxyService_DialNodeServer) error
+		return trace.Wrap(s.handleDialAuth(stream, r))
 	}
 
+	return trace.BadParameter("unknown dial request")
```
`frame.GetDialRequest()` will always be nil here. But perhaps a better error message would be something like `invalid first frame: expected DialRequest or DialAuth`

dboslee replied:
Updated here [b8539ad](https://github.com/gravitational/teleport/pull/24702/commits/b8539ad41d8d63d3d2a75034b62efcd8368b3b48)

dboslee committed:
Add godocs for DialAuth

dboslee committed:
Improve error messaging for invalid first frame

dboslee committed:
Improve proxy peering service logging

dboslee committed:
Update godocs for new cluster dialer behavior

dboslee committed:
Remove unnecessary context

dboslee committed:
Fix incorrect godoc

dboslee commented on lib/reversetunnel/remotesite.go:
```diff
@@ -898,13 +990,64 @@ func UseTunnel(logger *log.Entry, c *sshutils.ChConn) bool {
+		return nil, trace.Wrap(err)
+	}
+
+	s.logger.Info("Dialing remote cluster over peer proxy")
```
I was thinking a dial to the node or auth is considered dialing the remote cluster in some way or another, but for clarity we can be more explicit about what we are dialing

dboslee replied:
Updated here [f6d6f31](https://github.com/gravitational/teleport/pull/24702/commits/f6d6f3146e242a26ed09c7a2a8961f7322bb7103)

dboslee committed:
Update remote site debug logging

dboslee committed:
Update godocs

dboslee committed:
Add more detail to failed netconfig requests

dboslee commented on lib/proxy/peer/client.go:
```diff
@@ -339,6 +340,33 @@ func (c *Client) updateConnections(proxies []types.Server) error {
 	return trace.NewAggregate(errs...)
 }
 
+func (c *Client) DialAuth(clusterName string, src net.Addr, dst net.Addr, proxyIDs []string) (net.Conn, error) {
```
Added here [cfefce7](https://github.com/gravitational/teleport/pull/24702/commits/cfefce7755a9d9b1dd4f4bcc286603682a83176d)

dboslee commented on lib/proxy/peer/service.go:
```diff
@@ -40,28 +40,38 @@ func (s *proxyService) DialNode(stream proto.ProxyService_DialNodeServer) error
+		"src":     dial.Source.Addr,
+		"dst":     dial.Destination.Addr,
+	})
+	log.Debugf("Node dial request from peer.")
```
Updated here [4020bb5](https://github.com/gravitational/teleport/pull/24702/commits/4020bb5c98de7620075437babfba04dc21aae1aa)

dboslee commented on lib/proxy/peer/service.go:
```diff
@@ -116,6 +169,7 @@ func splitServerID(address string) (string, string, error) {
 // ClusterDialer dials a node in the given cluster.
```
Updated here [ac81c7e](https://github.com/gravitational/teleport/pull/24702/commits/ac81c7e78a4a33cd5bea151c95a902edf4acc8b5)

dboslee commented on lib/reversetunnel/agent.go:
```diff
@@ -592,6 +596,28 @@ func (a *agent) handleDrainChannels() error {
+	ctx, cancel := context.WithTimeout(ctx, defaults.DefaultIOTimeout)
+	defer cancel()
+
+	ok, payload, err := a.client.SendRequest(ctx, teleport.NetconfigRequest, true, nil)
```
removed context here [1ef38e1](https://github.com/gravitational/teleport/pull/24702/commits/1ef38e1f08338a751de30df9cd9d0a0209ebd588)

dboslee commented on lib/reversetunnel/agentpool.go:
```diff
@@ -569,89 +646,88 @@ func (p *AgentPool) transport(ctx context.Context, channel ssh.Channel, requests
-		return false
-	}
-	return c.tunnelStrategyType == types.ProxyPeering
+// reportConnectedProxies returns true if the number of agents should be restricted.
```
fixed here [2ecfcc8](https://github.com/gravitational/teleport/pull/24702/commits/2ecfcc85d1893147908b1b467cad2e660334e225)

dboslee commented on lib/srv/regular/sshserver.go:
```diff
@@ -1903,6 +1905,30 @@ func (s *Server) handleVersionRequest(req *ssh.Request) {
+	}
+	payload, err := json.Marshal(netconfig)
+	if err != nil {
+		req.Reply(true, []byte(err.Error()))
```
Updated here [16bf943](https://github.com/gravitational/teleport/pull/24702/commits/16bf9433c5242f2cb3410b71b5786acc9f423908)

dboslee committed:
Merge remote-tracking branch 'origin/master' into david/reversetunnel/remote-site

dboslee committed:
Fix spelling

dboslee committed:
Move netconfig global request handler

dboslee committed:
Move import to please the linter

espadolini commented on lib/reversetunnel/remotesite.go:
```diff
@@ -760,36 +857,40 @@ func (s *remoteSite) Dial(params DialParams) (net.Conn, error) {
 	// an in-memory forwarding server.
-	if shouldDialAndForward(params, recConfig) {
-		return s.dialAndForward(params)
+	if shouldDialAndForward(params, recConfig, false) {
```
```suggestion
	const isLocalSite = false
	if shouldDialAndForward(params, recConfig, isLocalSite) {
```

dboslee replied:
addressed here [4677c65](https://github.com/gravitational/teleport/pull/24702/commits/4677c657c0dbd42a3e15b20af50358edcfac2e3c)

espadolini commented on lib/reversetunnel/localsite.go:
```diff
@@ -246,7 +246,7 @@ func (s *localSite) Dial(params DialParams) (net.Conn, error) {
 	// requested or the target server is a registered OpenSSH node, build
 	// an in-memory forwarding server.
-	if shouldDialAndForward(params, recConfig) {
+	if shouldDialAndForward(params, recConfig, true) {
```
```suggestion
	const isLocalSite = true
	if shouldDialAndForward(params, recConfig, isLocalSite) {
```

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -636,6 +544,55 @@ func (s *server) Shutdown(ctx context.Context) error {
+	case teleport.NetconfigRequest:
+		s.handleNetConfigRequest(ctx, r)
+	default:
+		if r.WantReply {
```
`Reply` to a request that doesn't want a reply is a noop internally to x/crypto/ssh, so you can avoid this check and always `r.Reply(false, nil)`.

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -636,6 +544,55 @@ func (s *server) Shutdown(ctx context.Context) error {
+		return
+	}
+
+	s.log.Debugf("david: handling netconfig request")
```
`david`? 🤔

dboslee replied:
whoops some debug code I didn't remove

espadolini reviewed (commented):
I'd rather see the `ClusterNetworkingConfig` sent as part of the discovery messages, so the client side of the reverse tunnel doesn't need an extra roundtrip and extra fallible steps - it would also have the advantage of not needing to update the reverse tunnel "protocol" together with the rest of the changes.

dboslee commented on api/proto/teleport/legacy/client/proto/proxyservice.proto:
```diff
@@ -48,6 +49,16 @@ message DialRequest {
 }
 
+// DialAuth is a request to dial a remote clusters auth server.
+message DialAuth {
```
Proxies that are unaware of the DialAuth message type will see a nil `DialRequest` and reject the dial attempt. For that same reason I think its a good idea to use a separate message type than overload the `DialRequest`

dboslee commented on api/proto/teleport/legacy/client/proto/proxyservice.proto:
This seems like something that should be addressed separately

dboslee committed:
Open channel to ensure reverse tunnel server handles global requests

dboslee committed:
Clean up debug code

dboslee committed:
Remove unnecessary check to want reply

dboslee committed:
Add is local site const

dboslee commented:
Piggy backing on the discovery request turned out to be pretty complex when trying to implement it. A few things I ran into....
1. Do we include it on first request or every request?
2. The discovery requests are sent sometime after the agent connection is established so we aren't guaranteed  to have received the netconfig (unless we block) when we go to establish another agent connection.
3. Whether we send the netconfig on the first or every request we need some mechanism to receive these events and always return the last received netconfig.

2 is my biggest hangup on this approach because we either have to wait until we receive a discovery request to continue connecting agents or continue connecting agents without knowing the netconfig.

The adds a channel to check that global requests are handled and then sends a global request. Also note this only ever gets used by remote clusters.

dboslee committed:
Update logic to retry request on unexpected errors

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -636,6 +544,49 @@ func (s *server) Shutdown(ctx context.Context) error {
+	// that does not support the request and a transient error.
+	netconfig, err := s.localAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		req.Reply(true, []byte(fmt.Sprintf("failed to get cluster networking config: %v", err)))
+	}
+	payload, err := json.Marshal(netconfig)
+	if err != nil {
+		req.Reply(true, []byte(fmt.Sprintf("failed to marshal cluster networking config: %v", err)))
```
I really don't like relying on the fact that an error string doesn't unmarshal as a json object to do error detection on the other end. Can we use the `ok` bool of the reply to the request? The payload still goes through even if a request in NAKed.

dboslee replied:
What do you think of a wrapper type to contain the error like 
```golang
type NetconfigWithError struct {
  ClusterNetworkingConfig types.ClusterNetworkingConfig
  Error *string
}
```

This makes it clear the request was not rejected and allows the client to check if the error is not nil.

espadolini replied:
Why does it matter if the request was "not rejected"?

The agent itself replies with `Reply(r, true, []byte(version)` or `Reply(r, false, []byte("Failed to retrieve auth version"))` to a version request through a global request, I'd like to not add yet another standard for message passing through SSH.

dboslee replied:
I suppose its not absolutely necessary. However the point was to allow the agentpool to differentiate between the request failing and the request not being supported. If the request failed this time around that means we can retry later but if the request is rejected (or in other words not supported) that means the agent pool should hit the /webapi/find endpoint and cobble together a netconfig with some defaults like using  the `agent_mesh` tunnel strategy. This helps maintain backwards compatibility.

Currently when the agentpool fails to retrieve a new netconfig it stops trying to connect new agents until it is able to retrieve the netconfig. If we know the request is not supported we can fallback to the default netconfig and continue connecting agents, if we know the request failed server side and can be retried we can retry.

Technically we know that if we can open the reverse tunnel v2 channel we know the netconfig request is supported and can be retried but it feels a kind of brittle to make that assumption.

dboslee replied:
Updated here [4bfe263](https://github.com/gravitational/teleport/pull/24702/commits/4bfe2630fe272c559e66e63397846acec51aa2c6)

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -964,10 +936,35 @@ func (s *server) upsertRemoteCluster(conn net.Conn, sshConn *ssh.ServerConn) (*r
+	}
+	var err error
+	if site == nil {
+		site, err = newRemoteSiteWithPeerTunnels(s, clusterName, tunnels)
```
`newRemoteSiteWithPeerTunnels` ends up calling `newRemoteSite` which does a whole bunch of I/O, and we're under a lock here.

dboslee replied:
Still need to address this. I am planning to split out the construction of the remote site from connecting to the remote auth server. That way we can construct the remotesite and store it and connect in the background.

espadolini commented on lib/sshutils/server.go:
```diff
@@ -631,6 +631,8 @@ func (s *Server) HandleConnection(conn net.Conn) {
 			} else {
 				span.End()
+				// Reject all global requests if there is not a handler present.
+				req.Reply(false, nil)
```
As we're entering this brand new world where we don't violate the SSH spec, can I bother you with adding a couple of `go ssh.DiscardRequests(reqC)` for requests that are just ignored at the moment?

`chanDiscovery` requests are either ignored or received and never replied to (in `lib/reversetunnel/conn.go` and `lib/reversetunnel/agent.go`), `chanHeartbeat` is drained of requests but the requests are never replied to (in `lib/reversetunnel/agent.go` and in `lib/reversetunnel/localsite.go`/`lib/reversetunnel/remotesite.go`).

dboslee replied:
I would prefer to address that in a separate PR.

dboslee commented on lib/reversetunnel/srv.go:
```diff
@@ -625,6 +633,17 @@ func (s *server) HandleNewChan(ctx context.Context, ccx *sshutils.ConnectionCont
+// handleServerV2 accepts a channel from a client to indicate global requests
+// are handled by the server properly.
+func (s *server) handleServerV2(nch ssh.NewChannel) error {
+	ch, reqs, err := nch.Accept()
```
What code do you suggest using? Seems like we return `ssh.ConnectionFailed` for the default reject case. So something other than that would differentiate it from a normal rejection from the clients perspective.

dboslee replied:
Let me know what you think of this [4ac0f33](https://github.com/gravitational/teleport/pull/24702/commits/4ac0f3354a7e2fd84e929abac12cd7a93ff43c06)

espadolini replied:
I like it!

dboslee commented on lib/reversetunnel/agent.go:
```diff
@@ -592,6 +603,52 @@ func (a *agent) handleDrainChannels() error {
+// GetClusterNetworkConfig gets the cluster networking config from the connected proxy.
+// trace.NotImplemented is returned when the proxy rejects the request.
+func (a *agent) GetClusterNetworkConfig(ctx context.Context) (types.ClusterNetworkingConfig, error) {
+	a.serverVersionMu.Lock()
+	defer a.serverVersionMu.Unlock()
```
I want to hold the lock while we open the `ReverseTunnelServerV2Channel` channel so it's only done once. In theory I think the agent pool only ever calls this synchronously anyways. However I can refactor so the lock is released right after the channel is handled and before the netconfig request.

> Can we move the serverHandlesGlobalRequests check to right after the handshake while nothing else is going on, so we don't have to use a mutex at all?

We could but this is only needed for remote clusters so having it here saves us from opening the channel in most cases.

espadolini replied:
If we had that information handy we could start asking for a reply to the heartbeat message, to allow for tighter idle timeouts on the agent side of the connection.

espadolini replied:
(not in this PR, but in the future)

dboslee replied:
Updated in [8ccb235](https://github.com/gravitational/teleport/pull/24702/commits/8ccb235c19a3070d8c79e0fa21272b40252598a6)

espadolini replied:
We're still doing IO while holding a lock later in this function; if we assume we won't call `handlesGlobalRequests` a lot, we can just drop the lock before doing IO and acquire it again afterwards to write the results.

dboslee committed:
Always reject requests on error

This assumes that if the reverse tunnel v2 channel succeeds then
the netconfig request is supported. If this were not the case the
agent pool could get stuck indefinitely.

dboslee committed:
Change discard and close ordering

dboslee committed:
Reduce scope of lock and rename some things

dboslee committed:
Fix agent test

dboslee commented on lib/reversetunnel/agent.go:
```diff
@@ -592,6 +603,52 @@ func (a *agent) handleDrainChannels() error {
+	}
+
+	// Failing to unmarshal most likely indicates a server side network failure
+	// that can be retried.
```
Nope, this was the part relying on failed unmarshal to indicate a server side error. This behavior has been removed and so has the comment.

dboslee committed:
Reject v2 channel with special reject code

dboslee committed:
Update lib/reversetunnel/agent.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

dboslee committed:
Update error handling to use errors.As

dboslee commented on lib/reversetunnel/agent.go:
```diff
@@ -592,6 +602,63 @@ func (a *agent) handleDrainChannels() error {
+		// REMOVE IN V14: In v14 proxy servers will be guarenteed to properly handle
+		// global requests.
+		v2Chan, v2Reqs, err := a.client.OpenChannel(a.ctx, teleport.ReverseTunnelServerV2Channel, nil)
+		if err, ok := err.(*ssh.OpenChannelError); ok {
```
Figured it out here [db08e2e](https://github.com/gravitational/teleport/pull/24702/commits/db08e2e7afda4502ef123718d8b86bce704ecaca)

dboslee committed:
Fix typo

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

dboslee committed:
Resolve lint errors

dboslee committed:
Add remote client manager to handle connecting clients in the background

dboslee committed:
Fix lint errors

dboslee committed:
Add backoff and missing cert cache

dboslee committed:
Merge remote-tracking branch 'origin/master' into david/reversetunnel/remote-site

dboslee committed:
Remove debug log

dboslee committed:
Add license header

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -636,6 +544,49 @@ func (s *server) Shutdown(ctx context.Context) error {
+		return
+	}
+
+	// On error, reply with true to indicate that the server is able to handle the request
+	// but a server side error occurred, this lets the agent differentiate between a server
+	// that does not support the request and a transient error.
```
```suggestion
```

espadolini commented on lib/reversetunnel/srv.go:
```diff
@@ -636,6 +544,49 @@ func (s *server) Shutdown(ctx context.Context) error {
+	// On error, reply with true to indicate that the server is able to handle the request
+	// but a server side error occurred, this lets the agent differentiate between a server
+	// that does not support the request and a transient error.
+	netconfig, err := s.localAccessPoint.GetClusterNetworkingConfig(ctx)
```
Should we have an explicit (short!) timeout for this?

espadolini commented on lib/reversetunnel/remoteclientmanager.go:
```diff
@@ -0,0 +1,308 @@
+
+	m.wg.Add(1)
+	m.connectLock.Unlock()
+	go func() {
+		defer m.wg.Done()
+		errC <- m.connect(ctx)
+	}()
+	return trace.Wrap(<-errC)
```
This is a very strange way of doing `defer m.wg.Done(); return trace.Wrap(m.connect(ctx))`. Why do we need to bump and reduce the waitgroup?

espadolini commented on lib/reversetunnel/remoteclientmanager.go:
```diff
@@ -0,0 +1,308 @@
+		}
+
+		m.clientLock.Lock()
+		m.auth = client
+		m.accessPoint = accessPoint
+		m.nodeWatcher = nodeWatcher
+		m.caWatcher = caWatcher
```
This kinda begs to be moved into a dedicated `remoteClientManagerState` struct.

espadolini commented on lib/reversetunnel/remoteclientmanager.go:
```diff
@@ -0,0 +1,308 @@
+	}
+}
+
+// Connect blocks while all clients are created. If connect is called more than once
+// existing clients will be closed and new clients created. Connect will always fail
+// after Close is called.
```
What's the point of being able to `Connect` multiple times? We only call it once from `newRemoteSite`, and it makes things really confusing.

dboslee replied:
I started writing this thinking we would want to reconnect. But now realize the underlying clients should reconnect on their own. Let me refactor this and it should address some of your concerns on it being too complicated. 😄

espadolini commented on lib/reversetunnel/remoteclientmanager.go:
```diff
@@ -0,0 +1,308 @@
+		if accessPoint != nil {
+			accessPoint.Close()
+		}
+		accessPoint, err = m.newAccessPointFunc(ctx, client, version)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote access point.")
+			continue
+		}
+
+		if nodeWatcher != nil {
+			nodeWatcher.Close()
+		}
+		nodeWatcher, err = m.newNodeWatcherFunc(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote node watcher.")
+			continue
+		}
+
+		if caWatcher != nil {
+			caWatcher.Close()
+		}
+		caWatcher, err = m.newCAWatcher(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote CA watcher.")
+			continue
+		}
```
None of these get closed if later steps fail with an error.

espadolini reviewed (changes requested):
The `remoteClientManager` needs to be trimmed down and simplified, it's far too complicated and I can't tell if we're liable to deadlock with the two locks. The IO while holding the lock in `agent` is also a problem.

espadolini reviewed (changes requested):
edit: dupe

dboslee committed:
Simplify remote client manager
